core/hybrisadaptor.cpp

```diff
@@ -391,23 +402,55 @@ void HybrisManager::initManager()
 
 HybrisManager::~HybrisManager()
 {
-    cleanup();
-
-#ifdef USE_BINDER
-    if (m_serviceManager) {
-        gbinder_servicemanager_unref(m_serviceManager);
-        m_serviceManager = NULL;
-    }
-#endif
+    /* This is exectuted after exiting main() function.
+     * No actions that need core application, binder ipc,
+     * android hal libraries, etc should be made.
+     */
 }
 
 void HybrisManager::cleanup()
 {
+    /* Stop any sensors that are active
+     */
+
     sensordLogD() << "stop all sensors";
     foreach (HybrisAdaptor *adaptor, m_registeredAdaptors.values()) {
         adaptor->stopSensor();
     }
 
+    /* Stop reacting to async events
+     */
+
+#ifdef USE_BINDER
+    gbinder_remote_object_remove_handler(m_remote, m_deathId);
+    m_deathId = 0;
+
+    if (m_pollTransactId) {
+        gbinder_client_cancel(m_client, m_pollTransactId);
+        m_pollTransactId = 0;
+
+        // The above code just marks down pending POLL transaction as
+        // to be cancelled later on when handler thread gets woken up.
+        //
+        // If we are exiting right after cleanup(), that is never going
+        // to happen and gbinder_ipc_exit() cleanup code blocks sensorfwd
+        // exit indefinitely.
+        //
+        // As a workaround, make a dummy synchronous POLL transaction to
+        // wake up the handler thread and get the cancellation processed.
+
+        GBinderLocalRequest *req = gbinder_client_new_request2(m_client, POLL);
+        int32_t status = 0;
+        gbinder_local_request_append_int32(req, 0);
+        GBinderRemoteReply *reply = gbinder_client_transact_sync_reply(m_client, POLL, req, &status);
```

Comment:
If this POLL works the way you described, I think this could be reduced to

```c
gbinder_client_transact(m_client, POLL, 0, req, NULL, NULL, NULL);`
```
and let it complete asynchronously in the background. You're ignoring the result of the call anyway.

Reply:
Tried this. Sometimes it works, sometimes exit hangs. Whereas I've not yet seen the wait for reply and then ignore it -variant hang. So, possibly there is some kind of timing component to this.

Reply:
Based on discussion here -> I tweaked the explanatory comments a bit, hopefully it is more clear now